File: NextSpring/src/main/java/com/study/nextspring/domain/post/repository/PostRepositoryImpl.java
package com.study.nextspring.domain.post.repository;

import com.querydsl.core.types.Order;
import com.querydsl.core.types.OrderSpecifier;
import com.querydsl.core.types.dsl.PathBuilder;
import com.querydsl.jpa.impl.JPAQuery;
import com.querydsl.jpa.impl.JPAQueryFactory;
import com.study.nextspring.domain.member.entity.Member;
import com.study.nextspring.domain.post.entity.Post;
import com.study.nextspring.global.base.KwTypeV1;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.support.PageableExecutionUtils;
import com.querydsl.core.BooleanBuilder;
import static com.study.nextspring.domain.post.entity.QPost.post;
@RequiredArgsConstructor
public class PostRepositoryImpl implements PostRepositoryCustom {

    private final JPAQueryFactory jpaQueryFactory;

    @Override
    public Page<Post> findByKw(KwTypeV1 kwType, String kw, Member author, Boolean published, Boolean listed, Pageable pageable) {
        BooleanBuilder builder = new BooleanBuilder();

        if (author != null) builder.and(post.author.eq(author));
        if (published != null) builder.and(post.published.eq(published));
        if (listed != null) builder.and(post.listed.eq(listed));

        if (kw != null && !kw.isBlank()) {
            applyKeywordFilter(kwType, kw, builder);
        }

        JPAQuery<Post> postsQuery = createPostsQuery(builder);
        applySorting(pageable, postsQuery);

        // offset과 limit을 설정
        postsQuery.offset(pageable.getOffset()).limit(pageable.getPageSize());

        JPAQuery<Long> totalQuery = createTotalQuery(builder);
        return PageableExecutionUtils.getPage(postsQuery.fetch(), pageable, totalQuery::fetchOne);
    }
    private void applyKeywordFilter(KwTypeV1 kwType, String kw, BooleanBuilder builder) {
        switch (kwType) {
            case TITLE -> builder.and(post.title.containsIgnoreCase(kw));
            case BODY -> builder.and(post.content.containsIgnoreCase(kw));
            case NAME -> builder.and(post.author.nickname.containsIgnoreCase(kw));
            case ALL -> builder.and(
                    post.title.containsIgnoreCase(kw)
                            .or(post.content.containsIgnoreCase(kw))
                            .or(post.author.nickname.containsIgnoreCase(kw))
            );
        }
    }

    private JPAQuery<Post> createPostsQuery(BooleanBuilder builder) {
        return jpaQueryFactory.select(post).from(post).where(builder);
    }

    private void applySorting(Pageable pageable, JPAQuery<Post> postsQuery) {
        for (Sort.Order order : pageable.getSort()) {
            PathBuilder pathBuilder = new PathBuilder(post.getType(), post.getMetadata());
            postsQuery.orderBy(new OrderSpecifier(order.isAscending() ? Order.ASC : Order.DESC, pathBuilder.get(order.getProperty())));
        }
    }

    private JPAQuery<Long> createTotalQuery(BooleanBuilder builder) {
        return jpaQueryFactory.select(post.count()).from(post).where(builder);
    }
}

File: k_rstudy/src/main/kotlin/com/krstudy/kapi/domain/post/repository/PostRepositoryImpl.kt
package com.krstudy.kapi.domain.post.repository

import com.krstudy.kapi.domain.member.entity.Member
import com.krstudy.kapi.domain.post.repository.PostRepositoryCustom
import com.querydsl.core.types.Order
import com.querydsl.core.types.OrderSpecifier
import com.querydsl.core.types.dsl.BooleanExpression
import com.querydsl.core.types.dsl.PathBuilder
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.support.PageableExecutionUtils
import org.springframework.stereotype.Repository
import com.querydsl.core.types.dsl.*
import com.krstudy.kapi.domain.post.entity.*

@Repository
class PostRepositoryImpl(
    private val jpaQueryFactory: JPAQueryFactory
) : PostRepositoryCustom {

    override fun search(author: Member?, isPublished: Boolean?, kw: String, pageable: Pageable): Page<Post> {
        // 조건 생성
        var condition: BooleanExpression = QPost.post.isNotNull()

        // 검색 조건을 로그로 출력
        println("Searching with author: $author, isPublished: $isPublished, kw: $kw")

        // author가 null이 아닐 경우 조건 추가
        author?.let {
            condition = condition.and(QPost.post.author.eq(author))
        }

        // isPublished가 null이 아닐 경우 조건 추가
        isPublished?.let {
            condition = condition.and(QPost.post.isPublished.eq(it))
        }

        // kw가 비어있지 않을 경우 제목 또는 본문에서 검색
        if (kw.isNotBlank()) {
            condition = condition.and(
                QPost.post.title.containsIgnoreCase(kw)
                    .or(QPost.post.body.containsIgnoreCase(kw))
            )
        }

        // 쿼리 생성
        val postsQuery = jpaQueryFactory
            .selectFrom(QPost.post)
            .where(condition)

        // 페이지네이션 설정
        postsQuery
            .offset(pageable.offset)
            .limit(pageable.pageSize.toLong())

        // 전체 개수를 가져오기 위한 쿼리
        val totalQuery = jpaQueryFactory
            .select(QPost.post.count())
            .from(QPost.post)
            .where(condition)

        // 페이지 결과 반환
        return PageableExecutionUtils.getPage(postsQuery.fetch(), pageable, totalQuery::fetchCount)
    }
}